freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserSyncService.java

```diff
@@ -274,6 +292,51 @@ private SyncStatusDetail internalSynchronizeStack(Stack stack, UmsUsersState ums
         }
     }
 
+    private Optional<String> getOptionalAzureObjectId(List<CloudIdentity> cloudIdentities) {
+        List<CloudIdentity> azureCloudIdentities = cloudIdentities.stream()
+                .filter(cloudIdentity -> cloudIdentity.getCloudIdentityDomain().hasAzureCloudIdentityDomain())
+                .collect(Collectors.toList());
+        if (azureCloudIdentities.isEmpty()) {
+            return Optional.empty();
+        } else if (azureCloudIdentities.size() > 1) {
+            throw new IllegalStateException(String.format("List contains multiple azure cloud identities = %s", cloudIdentities));
+        } else {
+            String azureObjectId = Iterables.getOnlyElement(azureCloudIdentities).getCloudIdentityName().getAzureCloudIdentityName().getObjectId();
+            return Optional.of(azureObjectId);
+        }
+    }
+
+    private Map<String, String> getAzureObjectIdMap(Map<String, List<CloudIdentity>> cloudIdentityMapping) {
+        ImmutableMap.Builder<String, String> azureObjectIdMap = ImmutableMap.builder();
+        cloudIdentityMapping.forEach((key, cloudIdentities) -> {
+            Optional<String> azureObjectId = getOptionalAzureObjectId(cloudIdentities);
+            if (azureObjectId.isPresent()) {
+                azureObjectIdMap.put(key, azureObjectId.get());
+            }
+        });
+        return azureObjectIdMap.build();
+    }
+
+    private void syncAzureObjectIds(Stack stack, UmsUsersState umsUsersState, BiConsumer<String, String> warnings) {
+        String envCrn = stack.getEnvironmentCrn();
+        LOGGER.info("Syncing Azure Object IDs for environment {}", envCrn);
+
+        Map<String, String> userToAzureObjectIdMap = getAzureObjectIdMap(umsUsersState.getUserToCloudIdentityMap());
```

Comment:
this adds cloud identities for all users, regardless of if they have access to the environment. Is this desirable? I would think that you would only care about users that have access to the environment. Imagine if an account has 1000 user but the environment only has 100. You would be syncing the other 900 users cloud identities.

You can filter this map by the users that are in the umsUsersState.getUserState().getUsers()

Reply:
Good point, I've added the filtering part. Also I've moved most of this logic out of UserSyncService and into CloudIdentitySyncService.